### PR TITLE
New info api

### PR DESCRIPTION
**Pull request by DvirDukhan**
This PR introduces a new `VecSimInfoIterator` object to consume index info in a generic way. Each call to `next` yields a `VecSim_InfoField` struct that holds:
1. info field type - enum of string/uint64/int64/double
2. info field name - string
3. info field value

consuming the iterator should be like:
```c
VecSimInfoIterator *infoIter = VecSimIndex_InfoIterator(index);
while(VecSimInfoIterator_HasNextField(infoIter)) {
    VecSim_InfoField *infoFiled = VecSimInfoIterator_NextField(infoIter);
    RedisModule_ReplyWithCString(rmctx, infoField->fieldName);
    switch(infoField->fieldType) {
    case INFOFIELD_STRING:
         RedisModule_ReplyWithCString(rmctx, infoField->stringValue);
         break;
    case INFOFIELD_INT64:
         RedisModule_ReplyWithLongLong(rmctx, infoField->integerValue);
         break;
    case INFOFIELD_UINT64:
         RedisModule_ReplyWithLongLong(rmctx, infoField->uintegerValue);
         break;
    case INFOFIELD_DOUBLE:
         RedisModule_ReplyWithDouble(rmctx, infoField->stringValue);
         break;
    default:
         break;
    }
}
```

### CODE REVIEW

File: src/VecSim/utils/vec_utils.h [hunk 1]
```diff
@@ -12,8 +13,44 @@ struct CompareByFirst {
     }
 };
 
+struct VecSimCommonStrings {
```
Comment (alonre24) [inline src/VecSim/utils/vec_utils.h:16]: Do you think that we need a class for this? Shouldn't we just use `#define`?

File: src/VecSim/algorithms/brute_force/brute_force.cpp [hunk 2]
```diff
@@ -205,15 +205,44 @@ VecSimIndexInfo BruteForceIndex::info() {
...
     info.bfInfo.indexSize = this->count;
     info.bfInfo.blockSize = this->vectorBlockSize;
-    info.memory = this->allocator->getAllocationSize();
+    info.bfInfo.memory = this->allocator->getAllocationSize();
     return info;
 }
 
+VecSimInfoIterator *BruteForceIndex::infoIterator() {
+    VecSimIndexInfo info = this->info();
+    VecSimInfoIterator *infoIterator = new VecSimInfoIterator(7);
```
Comment (alonre24) [inline src/VecSim/algorithms/brute_force/brute_force.cpp:219]: Avoid magic numbers?

File: src/VecSim/algorithms/hnsw/hnswlib_c.cpp [hunk 3]
```diff
@@ -92,17 +92,58 @@ VecSimIndexInfo HNSWIndex::info() {
...
     info.hnswInfo.levels = this->hnsw.getMaxLevel();
     info.hnswInfo.entrypoint = this->hnsw.getEntryPointLabel();
-    info.memory = this->allocator->getAllocationSize();
+    info.hnswInfo.memory = this->allocator->getAllocationSize();
     return info;
 }
 
+VecSimInfoIterator *HNSWIndex::infoIterator() {
+    VecSimIndexInfo info = this->info();
+    VecSimInfoIterator *infoIterator = new VecSimInfoIterator(11);
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/hnswlib_c.cpp:110]: Avoid magic numbers?

File: tests/unit/test_bruteforce.cpp [hunk 4]
```diff
@@ -329,13 +330,108 @@ TEST_F(BruteForceTest, test_bf_info) {
...
     ASSERT_EQ(info.algo, VecSimAlgo_BF);
-    ASSERT_EQ(info.d, d);
+    ASSERT_EQ(info.bfInfo.dim, d);
     // User args
     ASSERT_EQ(info.bfInfo.blockSize, 1);
     ASSERT_EQ(info.bfInfo.indexSize, 0);
     VecSimIndex_Free(index);
 }
 
+void compareFlatIndexInfoToIterator(VecSimIndexInfo info, VecSimInfoIterator *infoIter) {
```
Comment (alonre24) [inline tests/unit/test_bruteforce.cpp:340]: Consider moving it to `test_utils` (or at least to the top of the file) for making the tests code cleaner..

File: tests/unit/test_bruteforce.cpp [hunk 5]
```diff
@@ -329,13 +330,108 @@ TEST_F(BruteForceTest, test_bf_info) {
...
+    float v[d];
+    for (size_t i = 0; i < d; i++) {
+        v[i] = (float)i;
+    }
+    VecSimIndex *index = VecSimIndex_New(&params);
+    VecSimIndexInfo info = VecSimIndex_Info(index);
+    VecSimInfoIterator *infoIter = VecSimIndex_InfoIterator(index);
+    ASSERT_EQ(1, info.bfInfo.blockSize);
+    ASSERT_EQ(0, info.bfInfo.indexSize);
+    compareFlatIndexInfoToIterator(info, infoIter);
```
Comment (alonre24) [inline tests/unit/test_bruteforce.cpp:413]: Why the ASSERT here? This is validated in `compareFlatIndexInfoToIterator` no?

File: tests/unit/test_hnswlib.cpp [hunk 6]
```diff
@@ -313,14 +314,137 @@ TEST_F(HNSWLibTest, test_hnsw_info) {
...
-    ASSERT_EQ(info.d, d);
+    ASSERT_EQ(info.hnswInfo.dim, d);
     // User args
     ASSERT_EQ(info.hnswInfo.efConstruction, 1000);
     ASSERT_EQ(info.hnswInfo.M, 200);
     ASSERT_EQ(info.hnswInfo.efRuntime, 500);
     VecSimIndex_Free(index);
 }
 
+void compareHNSWIndexInfoToIterator(VecSimIndexInfo info, VecSimInfoIterator *infoIter) {
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:325]: Consider moving it, same as for `compareBFIndexInfoToIterator`

File: tests/unit/test_hnswlib.cpp [hunk 7]
```diff
@@ -313,14 +314,137 @@ TEST_F(HNSWLibTest, test_hnsw_info) {
...
+    VecSimIndex *index = VecSimIndex_New(&params);
+    VecSimIndexInfo info = VecSimIndex_Info(index);
+    VecSimInfoIterator *infoIter = VecSimIndex_InfoIterator(index);
+    ASSERT_EQ(100, info.hnswInfo.M);
+    ASSERT_EQ(250, info.hnswInfo.efConstruction);
+    ASSERT_EQ(400, info.hnswInfo.efRuntime);
+    ASSERT_EQ(0, info.hnswInfo.indexSize);
+    ASSERT_EQ(-1, info.hnswInfo.levels);
+    ASSERT_EQ(-1, info.hnswInfo.entrypoint);
+    compareHNSWIndexInfoToIterator(info, infoIter);
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:424]: Why the ASSERT here, aren't you asserting it in `compareHNSWIndexInfoToIterator`?

File: tests/unit/test_hnswlib.cpp [hunk 8]
```diff
@@ -313,14 +314,137 @@ TEST_F(HNSWLibTest, test_hnsw_info) {
...
+            ASSERT_EQ(infoFiled->fieldType, INFOFIELD_UINT64);
+            ASSERT_EQ(infoFiled->uintegerValue, info.hnswInfo.efRuntime);
+        } else if (!strcmp(infoFiled->fieldName, VecSimCommonStrings::HNSW_M_STRING)) {
+            // M.
+            ASSERT_EQ(infoFiled->fieldType, INFOFIELD_UINT64);
+            ASSERT_EQ(infoFiled->uintegerValue, info.hnswInfo.M);
+        } else if (!strcmp(infoFiled->fieldName, VecSimCommonStrings::HNSW_LEVELS)) {
+            // Levels.
+            ASSERT_EQ(infoFiled->fieldType, INFOFIELD_UINT64);
+            ASSERT_EQ(infoFiled->uintegerValue, info.hnswInfo.levels);
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:365]: What is `levels`? The number of levels in HNSW ? Consider renaming to `max_level` or `level_num`...

**Review by alonre24 — COMMENTED** [review]
Nice and clean, few cosmetic comments

File: src/VecSim/utils/vec_utils.h (hunk 1 above)
Comment (DvirDukhan) [reply src/VecSim/utils/vec_utils.h:16]: I prefer to have a dedicated container for this and avoid `#define`. I think RediSearch also defines its own strings. I don't want to have collisions here.

File: tests/unit/test_hnswlib.cpp (hunk 7 above)
Comment (DvirDukhan) [reply tests/unit/test_hnswlib.cpp:424]: in this function, you compare the info data to the info iterator data. you don't check the data. This is just sanity.

**Review by alonre24 — APPROVED** [review]

**Pull request merged**